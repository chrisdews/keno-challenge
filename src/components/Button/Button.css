.action-button.active-button {
    border: 1px solid rgba(119, 187, 98, 0.774);
    background-color: green;
    font-size: 1rem;
  }
  
  .inactive-button {
    border: 1px solid grey;
    color: rgb(226, 84, 84);
  }

.action-button {
    width: 70px;
    height: 30px;
    margin: 0px 10px;
    background-color: white;
    border-radius: 3px;
  }
  
  .action-button:hover {
    background-color: rgba(119, 187, 98, 0.774);
  }
  
  .action-button.large {
      width: 130px;
  }
  
  .action-button.large.primary {
      background-color: rgba(64, 201, 219, 0.849);
      color: white;
  }
  
  .action-button.large.primary:hover {
      background-color: white;
      color: grey;
  }
  
  .large:hover {
      background-color: rgba(64, 201, 219, 0.849);
      color: white;
  }