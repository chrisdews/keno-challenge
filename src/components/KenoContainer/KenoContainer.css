.selected {
    background-color: aquamarine !important;
}

.min-width-button {
    min-width: 50px !important;
}