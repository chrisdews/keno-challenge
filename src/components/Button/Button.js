import React from 'react';
import PropTypes from 'prop-types';
import './Button.css'


const Button = ({children, className, onClick}) => {
    return (
        <button className={className} onClick={onClick}>
            {children}
        </button>
    );
};


Button.propTypes = {
    className: PropTypes.string,
    onClick: PropTypes.func,
    children: PropTypes.any
};


export default Button;
