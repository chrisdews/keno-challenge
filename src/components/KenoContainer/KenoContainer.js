import React, { useState, useEffect } from "react";
import "./KenoContainer.css";
import Header from "../Header";
import Button from "../Button"
import Grid from "@material-ui/core/Grid";

function KenoContainer() {
  const [selectedNumbers, setSelectedNumbers] = useState([]);
  const [stake, setStake] = useState(0);
  const [numbersConditionsMet, setNumbersConditionsMet] = useState(false);
  const [stakeConditionsMet, setStakeConditionsMet] = useState(false);
  const [gameMessage, setGameMessage] = useState(false);

  const titleContent = "Keno";
  const stakeButtonContent = [
    `${1} ðŸ’°`,
    `${2} ðŸ’°`,
    `${5} ðŸ’°`,
    `${10} ðŸ’°`,
    `${20} ðŸ’°`,
  ];

  const boardBuilder = (x, y) => {
    let board = [];
    for (let i = 0; i < y; i++) {
      let row = [];
      for (let j = 0; j < x; j++) {
        let num = j + 1 + x * i;

        let className = "";
        if (selectedNumbers.includes(num)) {
          className = "active";
        } else if (num <= x * y * 0.5) {
          className = "blue";
        } else {
          className = "red";
        }
        row.push(
          <Button
            key={`key ${num}`}
            onClick={() => {
              clickHandler(num);
            }}
            className={`${className} min-width-button number`}
          >
            {num}
          </Button>
        );
      }
      board.push(<Grid key={`row ${i + 1}`}>{row}</Grid>);
    }
    return board;
  };

  const clickHandler = (num) => {
    if (selectedNumbers.length > 4 && !selectedNumbers.includes(num)) {
      setSelectedNumbers(selectedNumbers.shift());
    }
    if (selectedNumbers.includes(num)) {
      setSelectedNumbers(selectedNumbers.filter((arrNum) => arrNum !== num));
    } else {
      setSelectedNumbers([...selectedNumbers, num]);
    }
  };

  const stakeButtonClickHandler = (num) => {
    setStake(num.split(" ")[0]);
  };

  const handleStakeInputChange = (event) => {
    setStake(event.target.value.split(" ")[0]);
  };

  const activeButton = (num) => {
    if (num.split(" ")[0] === stake) {
      return "active-button";
    } else {
      return "inactive-button";
    }
  };

  const luckyClickHandler = () => {
    let arr = [];
    for (let i = 0; i < 5; i++) {
      arr.push(Math.floor(Math.random() * Math.floor(80)));
    }
    setSelectedNumbers(arr);
  };

  const clearClickHandler = () => {
    setSelectedNumbers([]);
    setGameMessage(false);
    setNumbersConditionsMet(false);
    setStakeConditionsMet(false);
  };

  const submitClickHandler = () => {
    let isDecimal = stake.toString().split(".")[1];

    if (isDecimal && isDecimal.length > 2) {
      setGameMessage("too many decimals");
    } else if (!numbersConditionsMet) {
      setGameMessage("Select some numbers");
    } else if (!stakeConditionsMet) {
      setGameMessage("Choose a stake");
    } else if (numbersConditionsMet && stakeConditionsMet) {
      setGameMessage("You win!");
    } else {
      setGameMessage(false);
    }
  };

  useEffect(() => {
    if (selectedNumbers.length > 0) {
      setNumbersConditionsMet(true);
    } else {
      setNumbersConditionsMet(false);
    }
  }, [selectedNumbers]);

  useEffect(() => {
    if (stake > 0) {
      setStakeConditionsMet(true);
    } else {
      setStakeConditionsMet(false);
    }
  }, [stake]);

  return (
    <>
      <Header>{titleContent}</Header>
      <Grid>{boardBuilder(8, 10)}</Grid>
      <div className={"message-container"}>
        <div className={"message"}>{gameMessage && gameMessage}</div>
      </div>

      <div className="stake-container">
        {stakeButtonContent.map((num) => (
          <Button
            key={`buttons ${num}`}
            onClick={() => {
              stakeButtonClickHandler(num);
            }}
            className={"action-button " + activeButton(num)}
            
          >
            {num}
          </Button>
        ))}
        <input
          type="text"
          value={stake}
          onChange={(e) => {
            handleStakeInputChange(e);
          }}
        ></input>
      </div>

      <div className={"secondaryButtonContainer"}>
        <Button key="lucky" className="action-button large" onClick={luckyClickHandler}>Lucky Pick</Button>
        <Button key="clear" className="action-button large" onClick={clearClickHandler}>Clear Picks</Button>
      </div>

      <div className={"submitButtonContainer"}>
        <Button key="place" className="action-button large primary" onClick={submitClickHandler}>Place Bet</Button>
      </div>
    </>
  );
}

export default KenoContainer;
