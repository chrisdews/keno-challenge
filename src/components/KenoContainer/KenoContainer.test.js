import { array } from 'prop-types';
import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import KenoContainer from './KenoContainer';
import Button from '../Button'

const setup = () => {
  const kenoRender = render(<KenoContainer><Button></Button></KenoContainer>)
  const submitButton = kenoRender.getByRole('button', {name: 'Place Bet'})
  const promptNode = kenoRender.getByRole('prompt')
  const firstNumberButton = kenoRender.getByRole('button', {name: '1'})
  const setStakeButton = kenoRender.getByRole('button', {name: '1 ðŸ’°'})
  const stakeInput = screen.getByLabelText('stake-input')

  return {submitButton, promptNode, firstNumberButton, stakeInput, setStakeButton, ...kenoRender}
}

it('renders Snapshot correctly', () => {
  const tree = renderer
    .create(<KenoContainer />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Click Place Bet before selecting numbers renders prompt to select picks', () => {
  const {submitButton, promptNode} = setup()
  fireEvent.click(submitButton)
  expect(promptNode).toHaveTextContent(/select some numbers/i)
})

test('Click Place Bet before selecting a stake renders prompt to select a stake', () => {
  const {submitButton, promptNode, firstNumberButton} = setup()
  fireEvent.click(firstNumberButton)
  fireEvent.click(submitButton)
  expect(promptNode).toHaveTextContent(/choose a stake/i)
})

test('Click Place Bet after selecting numbers and stake renders win game message', async () => {
  const {submitButton, promptNode, firstNumberButton, setStakeButton} = setup()
  fireEvent.click(firstNumberButton)
  fireEvent.click(setStakeButton)
  fireEvent.click(submitButton)
  expect(promptNode).toHaveTextContent(/you win/i)
})

test('Click Place Bet after inputting stake with more than 2dp renders too many decimals prompt', () => {
  const {submitButton, promptNode, stakeInput} = setup()
  fireEvent.change(stakeInput, {target: {value: '23.456'}})
  fireEvent.click(submitButton)
  expect(promptNode).toHaveTextContent(/too many decimals/i)
})

test('Click Place Bet after inputting stake with less than 3dp and number selected renders win prompt', () => {
  const {firstNumberButton, submitButton, promptNode, stakeInput} = setup()
  fireEvent.click(firstNumberButton)
  fireEvent.change(stakeInput, {target: {value: '23.45'}})
  fireEvent.click(submitButton)
  expect(promptNode).toHaveTextContent(/you win/i)
})

test('There should never be more than 5 numbers selected', () => {
  const kenoRender = render(<KenoContainer><Button></Button></KenoContainer>)

  for (let index = 1; index < 10; index++) {
    const selectedNumber = kenoRender.getByRole('button', {name: `${index}`});
    fireEvent.click(selectedNumber)
  }

  const buttons = kenoRender.getAllByRole('button')
  let activeCount = 0
  buttons.forEach(element => {
    element.className.includes('active min-width-button') && activeCount++
  });
  expect(activeCount).toBeLessThan(6)
})
// this test is a bit weird, how would you do it? background colour doesnt seem to work for me.