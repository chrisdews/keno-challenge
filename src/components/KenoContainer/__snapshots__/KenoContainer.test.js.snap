// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Snapshot correctly 1`] = `
Array [
  <div
    className="header"
    data-testid="keno-header"
  >
    Keno
  </div>,
  <div
    className="MuiGrid-root"
  >
    <div
      className="MuiGrid-root"
    >
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        1
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        2
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        3
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        4
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        5
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        6
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        7
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        8
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        9
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        10
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        11
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        12
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        13
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        14
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        15
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        16
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        17
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        18
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        19
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        20
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        21
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        22
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        23
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        24
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        25
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        26
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        27
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        28
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        29
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        30
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        31
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        32
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        33
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        34
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        35
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        36
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        37
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        38
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        39
      </button>
      <button
        className="blue min-width-button number"
        onClick={[Function]}
      >
        40
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        41
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        42
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        43
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        44
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        45
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        46
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        47
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        48
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        49
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        50
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        51
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        52
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        53
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        54
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        55
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        56
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        57
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        58
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        59
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        60
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        61
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        62
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        63
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        64
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        65
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        66
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        67
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        68
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        69
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        70
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        71
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        72
      </button>
    </div>
    <div
      className="MuiGrid-root"
    >
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        73
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        74
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        75
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        76
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        77
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        78
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        79
      </button>
      <button
        className="red min-width-button number"
        onClick={[Function]}
      >
        80
      </button>
    </div>
  </div>,
  <div
    className="message-container"
  >
    <div
      className="message"
      role="prompt"
    />
  </div>,
  <div
    className="stake-container"
  >
    <button
      className="action-button inactive-button"
      onClick={[Function]}
    >
      1 💰
    </button>
    <button
      className="action-button inactive-button"
      onClick={[Function]}
    >
      2 💰
    </button>
    <button
      className="action-button inactive-button"
      onClick={[Function]}
    >
      5 💰
    </button>
    <button
      className="action-button inactive-button"
      onClick={[Function]}
    >
      10 💰
    </button>
    <button
      className="action-button inactive-button"
      onClick={[Function]}
    >
      20 💰
    </button>
    <input
      aria-label="stake-input"
      onChange={[Function]}
      type="text"
      value={0}
    />
  </div>,
  <div
    className="secondaryButtonContainer"
    data-testid="lucky-button-1"
  >
    <button
      className="action-button large"
      onClick={[Function]}
    >
      Lucky Pick
    </button>
    <button
      className="action-button large"
      onClick={[Function]}
    >
      Clear Picks
    </button>
  </div>,
  <div
    className="submitButtonContainer"
  >
    <button
      className="action-button large primary"
      onClick={[Function]}
    >
      Place Bet
    </button>
  </div>,
]
`;
